File: haskell/src/Fac.hs
module Fac
    ( factorials_seq
    , factorials_par
    ) where

import Control.Parallel.Strategies (rdeepseq, using, parList)

fac :: Integer -> Integer
fac 0 = 1
fac n = n * fac (n - 1)

factorials_seq :: [Integer] -> [Integer]
factorials_seq = map fac

factorials_par :: [Integer] -> [Integer]
factorials_par xs = map fac xs `using` parList rdeepseq

File: scala/src/main/scala/programs/Fac.scala
import scala.collection.parallel._
import scala.concurrent.forkjoin.ForkJoinPool
import scala.annotation.tailrec

package object fac {

def fac(n: Int): Int = {
  @tailrec
  def withAcc(acc: Int, x: Int): Int =
    if (x == 0) acc else withAcc(acc * x, x - 1)
  withAcc(1, n)
}

def factorials_seq(xs: Seq[Int]) = {
  xs.map(fac)
}

def factorials_par(xs: ParSeq[Int]) = {
  xs.map(fac)
}

}
